File: Data Structures/Java/CustomLinkedList.java
package com.mycompany.datastructures;

import java.util.*;

interface MyLinkedList {
	void insert(int index, Object value);

	void push(Object value);

	void delete(int index);

	Object get(int index);

	Object[] getAll();

	void set(int index, Object value);

	void reverse();

	int length();

}

public class CustomLinkedList implements MyLinkedList {
	private Node head;
	private int length = 0;

	public CustomLinkedList() {
		head = new Node(null, null);
	}

	public void push(Object value) {
		Node lastNode = getLastNode();
		Node newNode = new Node(value, null);
		lastNode.setNext(newNode);
		length++;
	}

	public void insert(int index, Object value) throws ArrayIndexOutOfBoundsException {
		// Our index sould be 0 <= index < length and length != 0
		if (length == 0) {
			throw new ArrayIndexOutOfBoundsException();
		}

		Node prevNode = getNode(index - 1);
		Node newNode = new Node(value, prevNode.getNext());
		prevNode.setNext(newNode);

	}

	private Node getNode(int index) throws ArrayIndexOutOfBoundsException {
		if (index < length && index >= 0) {
			int counter = 0;
			Node tmp = head;
			while (counter <= index) {
				tmp = tmp.getNext();
				counter++;
			}
			return tmp;
		} else if (index == -1) {
			return head;
		} else {
			throw new ArrayIndexOutOfBoundsException();
		}
	}

	private Node getLastNode() {
		Node tmp = head;
		while (tmp.getNext() != null) {
			tmp = tmp.getNext();
		}
		return tmp;
	}

	public Object get(int index) throws ArrayIndexOutOfBoundsException {
		return getNode(index).getData();
	}

	public Object[] getAll() {
		Object[] arr = new Object[length];
		int counter = 0;
		Node tmp = head;
		while (tmp.getNext() != null) {
			tmp = tmp.getNext();
			arr[counter] = tmp.getData();
			counter++;
		}
		return arr;
	}

	public void delete(int index) {
		Node prevNode = getNode(index - 1);
		Node currentNode = prevNode.getNext();
		prevNode.setNext(currentNode.getNext());
	}

	public void set(int index, Object value) {
		Node currentNode = getNode(index);
		currentNode.setData(value);
	}

	public void reverse() {
		if (head.getNext() == null) {
			return;
		}

		Node tmp = head.getNext();
		Node prev = null;
		Node next = tmp.getNext();

		while (next != null) {
			tmp.setNext(prev);
			prev = tmp;
			tmp = next;
			next = next.getNext();
		}
		tmp.setNext(prev);
		head.setNext(tmp);
	}

	public int length() {
		return length;
	}

	public void removeDuplicate() {
		HashSet<Integer> hashtable = new HashSet<Integer>();
		Node first = head.getNext();
		Node second = head;
		int data;
		while (first != null) {
			data = (int) first.getData();
			if (!hashtable.add(data)) {
				// duplication.. delete node
				second.setNext(first.getNext());
				first = first.getNext();
				length--;
				continue;
			}
			first = first.getNext();
			second = second.getNext();

		}
	}

	public Object[] toLast(int index) {
		Node tmp = head.getNext();
		Object[] result = new Object[length - index];
		int loopCounter = 0;
		int resultCounter = 0;

		while (tmp != null) {
			if (loopCounter >= index) {
				result[resultCounter] = tmp.getData();
				resultCounter++;
			}
			tmp = tmp.getNext();
			loopCounter++;
		}

		return result;
	}

	public Object getMiddle() {
		int index = (int) (length / 2);
		if (length <= 2) {
			return null;
		}
		Node node = getNode(index);
		return node.getData();

	}

	public static void main(String[] args) {
		CustomLinkedList test = new CustomLinkedList();
		test.push(5);
		test.push(3);
		test.push(4);
		test.push(5);
		test.push(10);

		Object[] values = test.getAll();
		for (Object element : values) {
			System.out.println(element.toString());
		}

		test.reverse();
		values = test.getAll();
		for (Object element : values) {
			System.out.println(element.toString());
		}

		test.removeDuplicate();
		values = test.getAll();
		for (Object element : values) {
			System.out.println(element.toString());
		}

		values = test.toLast(2);
		for (Object element : values) {
			System.out.println(element.toString());
		}

	}

}

File: Data Structures/Python/custom_linked_list.py
from typing import Protocol, Any, Union, List
from abc import abstractclassmethod
from dataclasses import dataclass


class MyLinkedList(Protocol):
    @abstractclassmethod
    def push(item: Any) -> None:
        ...

    @abstractclassmethod
    def insert(index: int, item: Any) -> None:
        ...

    @abstractclassmethod
    def pop() -> Any:
        ...

    @abstractclassmethod
    def delete(index: int) -> None:
        ...

    @abstractclassmethod
    def get(index: int) -> Any:
        ...

    @abstractclassmethod
    def set(index: int, item: Any) -> None:
        ...

    @abstractclassmethod
    def is_empty() -> bool:
        ...


@dataclass
class Node:
    data: Any
    next: "Node" = None


class CustomLinkedList(MyLinkedList):
    def __init__(self):
        self.__length = 0
        self.__head: Union[Node, None] = None

    def __len__(self):
        return self.__length

    def push(self, item: Any) -> None:
        # if list is empty than set self.__head to new node
        if (not self.__head):
            newNode = Node(item)
            self.__head = newNode
            self.__length += 1
            return

        # Go to last Node
        tmp = self.__head
        while (tmp.next):
            tmp = tmp.next

        newNode = Node(item)
        tmp.next = newNode
        self.__length += 1

    def is_empty(self) -> bool:
        if (self.__head == None):
            return True
        return False

    def __get_node(self, index: int) -> Node:
        if (index >= self.__length and index < 0):
            # Raise An Error
            ...

        counter = 0
        tmp = self.__head
        while (counter < index):
            tmp = tmp.next
            counter += 1

        return tmp

    def pop(self) -> Any:
        # Delete last node and return that node data
        if (self.is_empty()):
            return None

        if (self.__length == 1):
            data = self.__head.data
            self.__head = None
            self.__length -= 1
            return data

        prev_node = self.__get_node(self.__length - 2)
        current_node = prev_node.next
        prev_node.next = None
        self.__length -= 1

        return current_node.data

    def get(self, index: int) -> Any:
        if (self.is_empty):
            # Raise An Error
            ...
        return self.__get_node(index).data

    def set(self, index: int, item: Any) -> None:
        node = self.__get_node(index)
        node.data = item

    def insert(self, index: int, item: Any) -> None:
        # In order to insert, length of linked list has to be at least 1
        if (self.is_empty()):
            # Raise Error
            ...

        if (index == 0):
            newNode = Node(item, self.__head)
            self.__head = newNode
            self.__length += 1
            return

        prevNode = self.__get_node(index - 1)
        newNode = Node(item, prevNode.next)
        prevNode.next = newNode
        self.__length += 1

    def delete(self, index: int):
        if (self.is_empty()):
            # Raise Error
            ...

        if (index == 0):
            self.__head = self.__head.next
            self.__length -= 1

        prevNode = self.__get_node(index - 1)
        currentNode = prevNode.next
        prevNode.next = currentNode.next
        self.__length -= 1

    def get_all(self) -> List[Any]:
        itemList = []
        tmp = self.__head
        while (tmp):
            itemList.append(tmp.data)
            tmp = tmp.next
        return itemList


def main():
    test = CustomLinkedList()
    test.push("hello")
    test.push("world")
    arr = test.get_all()
    print(arr)

    test.insert(0, "insertion")
    arr = test.get_all()
    print(arr)

    test.delete(1)
    arr = test.get_all()
    print(arr)

    print(len(test))

    element = test.pop()
    arr = test.get_all()
    print(f"poped Element: {element}, LinkedList: {arr}, length: {len(test)}")


if (__name__ == "__main__"):
    main()
